event_scheduler/event_scheduler.py

```diff
@@ -18,28 +20,39 @@ def __init__(self, thread_name):
         super().__init__()
         self._scheduler_status_lock = threading.RLock()
         self._scheduler_status = self.SchedulerStatus.STOPPED
-        self._event_thread = threading.Thread(target=self.run, name=thread_name)
+        self._event_thread = threading.Thread(target=self.run,
+                                              name=thread_name)
+        # Condition variable to notify the event thread when there's a new
+        # event or when the deadline of the soonest event has passed.
+        self._cv = threading.Condition(self._lock)
+        # If we've looked at the front of the queue and the event isn't ready
+        # to execute, we set a timer for the remaining time. If a new event is
+        # added to the queue, then we cancel the timer and set it to None.
+        self._timer = None
 
-    def enterabs(self, time, priority, action, argument=(), kwargs={}):
+    def enterabs(self, time, priority, action, argument=(), kwargs=_sentinel):
         """Enter a new event in the queue at an absolute time.
         Returns an ID for the event which can be used to remove it,
         if necessary.
         """
+        if kwargs is _sentinel:
+            kwargs = {}
         with self._scheduler_status_lock:
             if self._scheduler_status != self.SchedulerStatus.RUNNING:
-                # TODO: Add error message
                 return None
-            super().enterabs(time, priority, action, argument, kwargs)
+            event = super().enterabs(time, priority, action, argument, kwargs)
+            # Notify the event thread about a new event in the queue.
+            with self._cv:
+                self._cv.notify()
```

Comment:
why is a lock required to notify?

Reply:
also you have a method `_notify` you made to do this

Reply:
That's just the way condition variables work. I think the thinking behind it is you notify the thread when the state of the locked resource has changed. You wouldn't want that resource changing again while you're doing the notify. In this case we're notifying whenever the state of the queue is changed (either when something is added to the queue or when it's time for the soonest event to execute). The _notify method exists so I can use the timer to call the notify function but you can't call it without the lock so I have to acquire it first.